File: android/src/main/java/com/emarsys/reactnative/utils/mappers/FilterMapper.kt
package com.emarsys.reactnative.utils.mappers

import com.emarsys.predict.api.model.RecommendationFilter
import com.emarsys.reactnative.utils.ArrayUtils.toStringList
import com.facebook.react.bridge.ReadableMap

object FilterMapper {

  fun toModel(readableMap: ReadableMap?): RecommendationFilter? {
    if (readableMap == null) {
      return null
    }

    val type = readableMap.getString("type")!!
    val field = readableMap.getString("field")!!
    val comparison = readableMap.getString("comparison")!!
    val expectations = readableMap.getArray("expectations")!!.toStringList()

    return when (type) {
      "INCLUDE" -> when (comparison) {
        "IS" -> RecommendationFilter.include(field).isValue(expectations[0])
        "IN" -> RecommendationFilter.include(field).inValues(expectations)
        "HAS" -> RecommendationFilter.include(field).hasValue(expectations[0])
        "OVERLAPS" -> RecommendationFilter.include(field).overlapsValues(expectations)
        else -> null
      }
      "EXCLUDE" -> when (comparison) {
        "IS" -> RecommendationFilter.exclude(field).isValue(expectations[0])
        "IN" -> RecommendationFilter.exclude(field).inValues(expectations)
        "HAS" -> RecommendationFilter.exclude(field).hasValue(expectations[0])
        "OVERLAPS" -> RecommendationFilter.exclude(field).overlapsValues(expectations)
        else -> null
      }
      else -> null
    }
  }

}

File: ios/utils/mappers/FilterMapper.m
#import "FilterMapper.h"

@implementation FilterMapper

static Mapper _toModel = ^EMSRecommendationFilter *(NSDictionary *dictionary) {
  if (dictionary == nil) {
    return nil;
  }

  NSString *type = [dictionary objectForKey:@"type"];
  NSString *field = [dictionary valueForKey:@"field"];
  NSString *comparison = [dictionary objectForKey:@"comparison"];
  NSArray *expectations = [dictionary objectForKey:@"expectations"];

  if ([type isEqualToString:@"INCLUDE"]) {
    if ([comparison isEqualToString:@"IS"]) {
      return [EMSRecommendationFilter includeFilterWithField:field isValue:expectations[0]];
    } else if ([comparison isEqualToString:@"IN"]) {
      return [EMSRecommendationFilter includeFilterWithField:field inValues:expectations];
    } else if ([comparison isEqualToString:@"HAS"]) {
      return [EMSRecommendationFilter includeFilterWithField:field hasValue:expectations[0]];
    } else if ([comparison isEqualToString:@"OVERLAPS"]) {
      return [EMSRecommendationFilter includeFilterWithField:field overlapsValues:expectations];
    }
  } else if ([type isEqualToString:@"EXCLUDE"]) {
    if ([comparison isEqualToString:@"IS"]) {
      return [EMSRecommendationFilter excludeFilterWithField:field isValue:expectations[0]];
    } else if ([comparison isEqualToString:@"IN"]) {
      return [EMSRecommendationFilter excludeFilterWithField:field inValues:expectations];
    } else if ([comparison isEqualToString:@"HAS"]) {
      return [EMSRecommendationFilter excludeFilterWithField:field hasValue:expectations[0]];
    } else if ([comparison isEqualToString:@"OVERLAPS"]) {
      return [EMSRecommendationFilter excludeFilterWithField:field overlapsValues:expectations];
    }
  }

  return nil;
};

+ (Mapper)toModel {
  return _toModel;
}

@end
